### PR TITLE
Add range request checks #105

### PR DESCRIPTION
**Pull request by brendanheywood**
#105

To test in web:

http://moodle.local/report/performance/index.php?detail=tool_heartbeat_rangerequestcheck

Or test new endpoint:

http://moodle.local/admin/tool/heartbeat/rangerequestcheck.php

or in cli:

```
$ php admin/cli/checks.php -t=performance -f=range -v
OK: All 'performance' checks OK
   Status | Check                                                       
----------+--------------------------------------------------------------------
       OK | Range requests check (tool_heartbeat_rangerequestcheck)     
          |     Range requests are working, 206 response with only 10 bytes
          |     of data
          |                                                             

```

### CODE REVIEW

File: classes/check/rangerequestcheck.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        curl_setopt($curl, CURLOPT_RANGE, '0-9');
+        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+
+        $response = curl_exec($curl);
+        $datalength = strlen($response);
+        $info = curl_getinfo($curl);
+
+        if ($response !== false) {
+            if ($info['http_code'] === 206 && $info['size_download'] == 10) {
+                // send_good("Range requests are working, 206 response with only 10 bytes of data");
```
Comment (Peterburnett) [inline classes/check/rangerequestcheck.php:56]: leftover here

File: rangerequestcheck.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+require_once(__DIR__.'/nagios.php');
+
+header("Content-Type: text/plain");
+
+// Make sure varnish doesn't cache this. But it still might so go check it!
+header('Pragma: no-cache');
+header('Cache-Control: private, no-cache, no-store, max-age=0, must-revalidate, proxy-revalidate');
+header('Expires: Tue, 04 Sep 2012 05:32:29 GMT');
+
+$check = new \tool_heartbeat\check\rangerequestcheck();
```
Comment (Peterburnett) [inline rangerequestcheck.php:37]: Will this explode on legacy? No API is no bueno

**Comment by Peterburnett** [comment]
I wonder if there is a metric we could define for confirming the response time, and seeing if it drops vs stays the same for serving. Might get real messy, but I wonder if we could diagnose the repeated full file pulls to CDN to serve the request before it becomes a bigger problem. Sounds like a too hard basket but worth thinking about

File: rangerequestcheck.php (hunk 2 above)
Comment (brendanheywood) [reply rangerequestcheck.php:37]: Yes but it's a new endpoint so I don't care about backwards compat, nothing will be calling it

**Comment by brendanheywood** [comment]
I think it is ok being black and white to start with, and they will all need some work to diagnose whats going on

**Pull request merged**